hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala

```diff
@@ -19,86 +19,152 @@ object TableStage {
+          bindIR(right.partition(GetField(ctxRef, rightCtxStructField))) { rPart =>
+            body(lPart, rPart)
+          }
+        }
+      })
   }
 
   def mapContexts(f: IR => IR): TableStage = {
-    val outer = this
-    new TableStage(letBindings, broadcastVals, globals, partitioner, f(contexts)) {
-      override def partition(ctxRef: Ref): IR = outer.partition(ctxRef)
-    }
+    val newContexts = f(contexts)
+    assert(newContexts.typ == contexts.typ)
+
+    new TableStage(letBindings, broadcastVals, globals, partitioner, newContexts, ctxRefName, partitionIR)
   }
 
   def mapGlobals(f: IR => IR): TableStage = {
     val newGlobals = f(globals)
-    val newID = genUID()
-    val outer = this
+    val globalsRef = Ref(genUID(), newGlobals.typ)
+
     new TableStage(
-      letBindings :+ newID -> newGlobals,
-      broadcastVals + newID,
-      Ref(newID, newGlobals.typ),
-      partitioner, contexts) {
-      def partition(ctxRef: Ref): IR = outer.partition(ctxRef)
-    }
+      letBindings :+ globalsRef.name -> newGlobals,
+      broadcastVals :+ globalsRef.name -> globalsRef,
+      globalsRef,
+      partitioner,
+      contexts,
+      ctxRefName,
+      partitionIR)
   }
 
-  def collect(bind: Boolean = true): IR = {
-    val ctx = Ref(genUID(), types.coerce[TStream](contexts.typ).elementType)
-    val broadcastRefs = MakeStruct(letBindings.filter { case (name, _) => broadcastVals.contains(name) })
+  def mapCollect(f: IR => IR): IR = {
+    val broadcastRefs = MakeStruct(broadcastVals)
     val glob = Ref(genUID(), broadcastRefs.typ)
-    val cda = CollectDistributedArray(contexts, broadcastRefs,
-      ctx.name, glob.name,
-      broadcastVals.foldLeft(partition(ctx))((accum, name) => Let(name, GetField(glob, name), accum)))
-    if (bind) wrapInBindings(cda) else cda
-  }
 
-  def collectWithGlobals(): IR = {
-    val mapped = mapPartition(ToArray)
-    mapped.wrapInBindings(
-      MakeStruct(FastIndexedSeq(
-        "rows" -> ToArray(flatMapIR(ToStream(mapped.collect(bind = false))) { elt => ToStream(elt) }),
-        "global" -> mapped.globals)))
+    CollectDistributedArray(
+      contexts, broadcastRefs,
+      ctxRefName, glob.name,
+      broadcastVals.foldLeft(f(partitionIR)) { case (accum, (name, _)) =>
+        Let(name, GetField(glob, name), accum)
+      })
   }
 
-  def changePartitionerNoRepartition(newPartitioner: RVDPartitioner): TableStage = {
-    new TableStage(letBindings, broadcastVals, globals, newPartitioner, contexts) {
-      def partition(ctxRef: Ref): IR = self.partition(ctxRef)
-    }
-  }
+  def collectWithGlobals(): IR =
+    MakeStruct(FastSeq(
+      "rows" -> ToArray(flatMapIR(ToStream(mapCollect(ToArray)))(ToStream)),
+      "global" -> globals))
+
+  def changePartitionerNoRepartition(newPartitioner: RVDPartitioner): TableStage =
+    new TableStage(letBindings, broadcastVals, globals, newPartitioner, contexts, ctxRefName, partitionIR)
```

Comment:
Could just do  `this.copy(rvdPartitioner = newPartitioner)` here I think, slightly less noisy.

Reply:
It could if `TableStage` was a Case class, or if we manually define a `copy` method. Now that there's no abstract `partition` method, we could do that, and you're right that would tidy up some things.